### DISCUSSION TITLE
Generated columns

### DISCUSSION THREAD

**Discussion by laurent512** [prisma/prisma#8891] [Q&A] [answered]
Hello,

Sorry in advance if question is basic, I am a beginner.

Is there a way, in the schema.prisma definition, to create a column which is always the result of a specific operation of other columns ? 

e.g. column 1 = Sum ( column 2 + column 3) ?

So that if column 3 is modified, column 1 will automatically be updated as well.

Thank you in advance

**Comment by ryands17** [comment, marked as answer]
@laurent512 👋 

If you're using Postgres, you can use it's feature of [generated columns](https://www.postgresql.org/docs/12/ddl-generated-columns.html) with Prisma Migrate.

For e.g.

Create a schema:

```prisma
model User {
  id        Int       @id @default(autoincrement())
  firstName String
  lastName  String?
  fullName  String?
}
```

Run `prisma migrate dev --create-only`

Edit the `.sql` file to use `generated`:

```sql
-- CreateTable
CREATE TABLE "User" (
    "id" SERIAL NOT NULL,
    "firstName" TEXT NOT NULL,
    "lastName" TEXT,
    "fullName" TEXT GENERATED ALWAYS AS ("firstName" || ' ' || "lastName") STORED,

    CONSTRAINT "User_pkey" PRIMARY KEY ("id")
);
```

Run `prisma migrate dev`.

Now you will get the result of `fullName` always generated.

**Reply by laurent512** [reply]
Thank you so much.

Would you mind to redirect me towards some documentation to write complex logic ?
In my case it would not be as simple as a sum or "AS ("firstName" || ' ' || "lastName") "
It would be true if ... or false otherwise

Thank you in advance :)

**Reply by ryands17** [reply]
Depends on the logic you want to write. Your logic should depend only on the fields and constant values and not any external parameters.

**Reply by laurent512** [reply]
Basically it is : 
Column A = true if (column B !=null && Column C==true && Column D > Today) ; false otherwise

(column D contains an expiry date in timestamp form)

**Reply by ryands17** [reply]
For complex conditions, you're better off defining a trigger that updates this as you create/update the record. Generated fields only accept immutable expressions and the above is not immutable.

**Reply by ryands17** [reply]
You can create a trigger in your migration `.sql` file and that should make this work.

**Reply by tsongas** [reply]
@ryands17 I've found this solution is not workable in terms of type safety, at least with my limited knowledge of TS. If I make `fullName` nullable as shown above, then I can't make it required in my GraphQL schema because Prisma says it could be null, which is inaccurate because the generated column always has a value. If I make `fullName` not null, then Prisma requires it to be passed in, which defeats the purpose of a generated column.

Any ideas?

**Reply by Bennison-Devadoss** [reply]
Prisma provides the preview option for full-text search (We can perform a full-text search using Prisma), Then how they don't have the features to define the generated column in the schema.prisma file?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
